### PR TITLE
Add support of AES-GCM to PKCS11 TA (take 2)

### PR DESCRIPTION
**Pull request by etienne-lms**
This P-R supersedes #6467.

### CODE REVIEW

File: ta/pkcs11/src/processing_symm.c [hunk 1]
```diff
@@ -263,6 +266,20 @@ allocate_tee_operation(struct pkcs11_session *session,
 	if (res == TEE_ERROR_NOT_SUPPORTED)
 		return PKCS11_CKR_MECHANISM_INVALID;
 
+	if (res == TEE_SUCCESS || params->id == PKCS11_CKM_AES_GCM) {
+		/*
+		 * Allocate a 2nd operation handler to save the operation state
+		 * on AES GCM one-short processing that queries the output
```
Comment (maroueneboubakri) [inline ta/pkcs11/src/processing_symm.c:272]: Typo: shot*

File: ta/pkcs11/src/processing_aes.c [hunk 2]
```diff
@@ -14,6 +15,410 @@
...
+		}
+
+		res = TEE_AEDecryptFinal(session->processing->tee_op_handle,
+					 NULL, 0, data_ptr, &data_size,
+					 ctx->pending_tag, ctx->tag_byte_len);
+
+		if (!data_size) {
+			TEE_Free(data_ptr);
+			data_ptr = NULL;
+			DMSG("Is this expected from the Core API?");
```
Comment (jforissier) [inline ta/pkcs11/src/processing_aes.c:339]: Should not happen indeed. I believe `assert(data_size);` would be better here.

File: ta/pkcs11/src/processing_symm.c [hunk 3]
```diff
@@ -963,6 +1024,43 @@ enum pkcs11_rc step_symm_operation(struct pkcs11_session *session,
...
+			rc = tee_ae_encrypt_final(session, out_buf,
+						  &ae_out_size);
+			output_data = true;
+			if (step == PKCS11_FUNC_STEP_ONESHOT)
+				out_size += ae_out_size;
+			else
+				out_size = ae_out_size;
+			break;
+		case PKCS11_FUNCTION_DECRYPT:
+			/* Now we're ready to releave data */
```
Comment (jforissier) [inline ta/pkcs11/src/processing_symm.c:1039]: s/releave/release/

**Review by jforissier — APPROVED** [review]
- For "ta: pkcs11: pkcs11_ta.h: define PKCS11_CKM_AES_GCM":
  `Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`
- For "ta: pkcs11: implement AES GCM operations":
  With minor comments
  `Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: ta/pkcs11/src/processing_symm.c (hunk 3 above)
Comment (maroueneboubakri) [reply ta/pkcs11/src/processing_symm.c:1039]: I believe should be reveal.

Comment (etienne-lms) [reply ta/pkcs11/src/processing_symm.c:1039]: indeed, thanks

**Comment by etienne-lms** [comment]
Comments addressed.
I've added a commit to the series: "ta: pkcs11: factorize second operation handle".

File: ta/pkcs11/src/processing_aes.c [hunk 4]
```diff
@@ -332,12 +332,7 @@ enum pkcs11_rc tee_ae_decrypt_final(struct pkcs11_session *session,
 		res = TEE_AEDecryptFinal(session->processing->tee_op_handle,
 					 NULL, 0, data_ptr, &data_size,
 					 ctx->pending_tag, ctx->tag_byte_len);
-
-		if (!data_size) {
-			TEE_Free(data_ptr);
-			data_ptr = NULL;
-			DMSG("Is this expected from the Core API?");
-		}
+		assert(!res && data_size);
```
Comment (jforissier) [inline ta/pkcs11/src/processing_aes.c:335]: Asserting `!res` may be a bit much, runtime errors can happen. However `data_size` should never be set to zero on output so `assert(data_size)` does make sense (but may not be that useful).

**Comment by jforissier** [comment]
For "ta: pkcs11: factorize second operation handle":

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: ta/pkcs11/src/processing_aes.c (hunk 4 above)
Comment (etienne-lms) [reply ta/pkcs11/src/processing_aes.c:335]: Ok, i'll fix that. Thanks.

**Comment by etienne-lms** [comment]
Review comment addressed.
@maroueneboubakri, if you're fine with this change, could you post a review tag?
@vesajaaskelainen, maybe would you like to have a look at this change?

**Comment by maroueneboubakri** [comment]
`Reviewed-by: Marouene Boubakri <marouene.boubakri@nxp.com>`

**Comment by etienne-lms** [comment]
I've squashed the fixup commits, applied the review tags and rebased the series.
Thanks for the reviews.

@jforissier, I've applied your A-b tag on commit "ta: pkcs11: implement AES GCM operations", despite I did not ask you if the fixup commit for https://github.com/OP-TEE/optee_os/pull/6683#pullrequestreview-1882811641 was fine. Please tell me if it's not ok.

**Comment by etienne-lms** [comment]
I would need review tags on the optee_test counter part: https://github.com/OP-TEE/optee_test/pull/730.

**Comment by jforissier** [comment]
@etienne-lms sure!

**Pull request merged**